alf/algorithms/planning_algorithm.py

```diff
@@ -269,3 +283,116 @@ def _calc_cost_for_action_sequence(self, time_step: TimeStep, state,
+                 elite_size=50,
+                 max_iter_num=5,
+                 epsilon=0.01,
+                 tau=0.9,
+                 scalar_var=None,
+                 upper_bound=None,
+                 lower_bound=None,
+                 name="CEMPlanAlgorithm"):
+        """Create a CEMPlanAlgorithm.
+
+        Args:
+            population_size (int): the size of polulation for optimization
+            planning_horizon (int): planning horizon in terms of time steps
+            elite_size (int): the number of elites selected in each round
+            max_iter_num (int|Tensor): the maximum number of CEM iterations
+            epsilon (float): a minimum variance threshold. If the variance of
+                the population falls below it, the CEM iteration will stop.
+            tau (float): a value in (0, 1) for softly updating the population
+                mean and variance:
+                    mean = (1 - tau) * mean + tau * new_mean
+                    var = (1 - tau) * var + tau * new_var
+            scalar_var (None|float): the value that will be used to construct
+                the initial diagonal covariance matrix of the multi-dimensional
+                Gaussian used by the CEM optimizer. If value is None,
+                0.5 * (upper_bound - lower_bound) is used.
+            upper_bound (int): upper bound for elements in solution;
+                action_spec.maximum will be used if not specified
+            lower_bound (int): lower bound for elements in solution;
+                action_spec.minimum will be used if not specified
+        """
+        super().__init__(
+            feature_spec=feature_spec,
+            action_spec=action_spec,
+            planning_horizon=planning_horizon,
+            upper_bound=upper_bound,
+            lower_bound=lower_bound,
+            name=name)
+
+        solution_size = planning_horizon * self._num_actions
+
+        self._plan_optimizer = CEMOptimizer(
+            solution_size,
+            self._population_size,
+            upper_bound=self._upper_bound,
+            lower_bound=self._lower_bound,
+            elite_size=elite_size,
+            max_iter_num=max_iter_num,
+            epsilon=epsilon,
+            tau=tau)
+
+        if scalar_var is None:
+            self._scalar_var = (self._upper_bound - self._lower_bound) / 2.
+        else:
+            self._scalar_var = scalar_var
+
+    def generate_plan(self, time_step: TimeStep, state, epsilon_greedy):
+        assert self._reward_func is not None, ("specify reward function "
+                                               "before planning")
+
+        assert self._dynamics_func is not None, ("specify dynamics function "
+                                                 "before planning")
+
+        batch_size = time_step.observation.shape[0]
+
+        self._plan_optimizer.set_cost(self._calc_cost_for_action_sequence)
+
+        prev_plan = state.planner.prev_plan
```

Comment:
one problem of this is that the first prev_plan is all zero. Will get a bad init_mean from it.

Reply:
Now add back the self-defined solution resetting which is more adaptive to solution bounds.